reconcile/vault_replication.py

```diff
@@ -237,7 +237,10 @@ def get_policy_secret_list(
         # Remove the * at the end of the path because list method expects
         # a folder path without any secret or wilcard
         path = path[:-1] if path.endswith("*") else path
-        secret_list.extend(vault_instance.list_all(path))
+        if path.endswith("/"):
+            secret_list.extend(vault_instance.list_all(path))
+        else:
+            secret_list.extend([path])
```

Comment:
or `.append(path)`?

Reply:
according comments above about `*`, we have 2 valid cases for `path`, either ends with `/*` or `/` as dir, or a single secret. Invalid case is like `policy/path/3/secret1*`


```python
if path.endswith("/*") or path.endswith("/"):
    secret_list.extend(vault_instance.list_all(path.rstrip("*")))
elif path.endswith("*"):
    raise ...
else:
    secret_list.append(path)
```

Reply:
Taking a look into the policies [documentation](https://developer.hashicorp.com/vault/docs/concepts/policies), the situation is a bit worse, since we can only properly support to replicate paths like

* path/to/file
* path/to/folder/*

but there are more combinations with `*` and `+`

I'll rework a bit the whole thing so that we fail properly.

Reply:
reworked in f36b21b763eb331df0298aacf6db744cb107279b

Reply:
that regex has an issue `/-_` means matches a single character in the range between / (index 47) and _ (index 95) (case sensitive). So this can match `policy/path/3;/secret1` I think what you mean is `^[A-Za-z0-9/\-_]+(?:/\*)?$` or just `^[\w/-]+(?:/\*)?$`, you can debug it at https://regex101.com/r/pp23Tm/1.

Another improvement is to move `re.compile` to a module const to avoid recompile every time the function got called.

And since we are already using regex to match the path, we can use match result directly to avoid recheck ends with `/*` or `/`, note we need non greedy for the first part, otherwise the last `/` will be captured by first part. And named capture for readability. Like this:

```python

SECRET_PATH_PATTERN = re.compile(r"^[\w/-]+?(?P<folder>/\*?)?$")


match = SECRET_PATH_PATTERN.match(path)
if not match:
  raise

if match.group("folder"):
  secrets.update(vault_instance.list_all(path.rstrip("*")))
else:
  secrets.add(path)
```

Reply:
nice catch! Fixed in 533875ee0dc3eb459c008d164973231dc9696c0c